flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -108,18 +109,21 @@ public Boundedness getBoundedness() {
         return sources.get(sources.size() - 1).boundedness;
     }
 
+    @Internal
```

Comment:
it seems strange that overridden methods have such an annotation. Shouldn't that be entirely defined by the interface/class we implement/extend?

Reply:
I think one important piece that is currently missing in Flink's annotation is the difference between **usage** and **implementations**. From a base interface perspective, connector developers are supposed to implement the method therefore it is not internal but from a user perspective, it is internal because no user should ever interact with the method.

Reply:
That is indeed something that we're missing. We discussed it a while ago and couldn't come up with a definitive answer.

I'd like to avoid trying to work around it this way.
In practice you can't make a breaking change to this method anyway because the method is still public as defined by the interface (so we don't gain any flexibility to change stuff) nor can you make any assumption that this method is not called because hey, a user might just work against the Source interface where this method is public.
So...I don't see us gaining anything really.

Reply:
> We did this approach already for a lot of connectors

Do we? Can you point me to some discussion where we agreed to do this?

> the connector gives no guarantee about maintaining the compatibility i.e. the returned serializer changes.

But the connector can already change whatever it returns so long as it adheres to the interface. I could see issues where the return type is a connector-specific implementation, although then I wonder whether this is actually correct / necessary.

Reply:
Good god it actually seems necessary; e.g., KafkaSink#createWrite actually implements 2 interfaces at once 🤯.

Reply:
I guess I'm fine with it then, but I still think it's a bit strange and would've rather pushed for a proper distinction of who consumes a given method.

Reply:
> Good god it actually seems necessary; e.g., KafkaSink#createWriter actually implements 2 distinct interfaces at once 🤯.

That was one of the design decisions for the new interface. That connector developer can choose which methods to implement and make it transparent for the end-user. Indeed that reduces a bit the readability.